pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,171 @@
+        timer.start_time("validate")
+        timer.end_time("test")
+    """
+
+    def __init__(
+        self,
+        duration: Optional[Union[str, timedelta, Dict[str, int]]] = None,
+        interval: str = Interval.step,
+        verbose: bool = True,
+    ) -> None:
+        super().__init__()
+        if isinstance(duration, str):
+            dhms = duration.strip().split(":")
+            dhms = [int(i) for i in dhms]
+            duration = timedelta(days=dhms[0], hours=dhms[1], minutes=dhms[2], seconds=dhms[3])
+        if isinstance(duration, dict):
+            duration = timedelta(**duration)
+        if interval not in set(Interval):
+            raise MisconfigurationException(
+                f"Unsupported parameter value `Timer(interval={interval})`. Possible choices are:"
+                f" {', '.join(set(Interval))}"
+            )
+        self._duration = duration
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = {stage: None for stage in RunningStage}
+        self._end_time = {stage: None for stage in RunningStage}
+        self._offset = timedelta()
+
+    def start_time(self, stage: str = RunningStage.TRAINING) -> Optional[datetime]:
+        """Return the start time of a particular stage"""
+        stage = RunningStage(stage)
+        return self._start_time[stage]
+
+    def end_time(self, stage: str = RunningStage.TRAINING) -> Optional[datetime]:
+        """Return the end time of a particular stage"""
+        stage = RunningStage(stage)
+        return self._end_time[stage]
+
+    def time_elapsed(self, stage: str = RunningStage.TRAINING) -> timedelta:
+        """Return the time elapsed for a particular stage"""
+        start = self.start_time(stage)
+        end = self.end_time(stage)
+        offset = self._offset if stage == RunningStage.TRAINING else timedelta(0)
+        if start is None:
+            return offset
+        if end is None:
+            return datetime.now() - start + offset
+        return end - start + offset
+
+    def time_remaining(self, stage: str = RunningStage.TRAINING) -> Optional[timedelta]:
+        if self._duration is not None:
+            return self._duration - self.time_elapsed(stage)
+
+    def on_train_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.TRAINING] = datetime.now()
+
+    def on_train_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.TRAINING] = datetime.now()
+
+    def on_validation_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.VALIDATING] = datetime.now()
+
+    def on_validation_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.VALIDATING] = datetime.now()
+
+    def on_test_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.TESTING] = datetime.now()
+
+    def on_test_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.TESTING] = datetime.now()
```

Comment:
n00b q: would this fail with daylight savings time? should we use t`ime.monotonic()` in case system clocks are reset or rewound?

Reply:
I switched everything to monotonic, but note now time_elapsed() etc returns seconds.
Hope that's fine.